wallet-core/src/lib.rs

```diff
@@ -22,9 +22,35 @@ pub const RNG_SEED: usize = 64;
 // phoenix-transaction
 const MAX_INPUT_NOTES: usize = 4;
 
+use alloc::collections::btree_map::BTreeMap;
 use alloc::vec::Vec;
 
-use execution_core::transfer::phoenix::{Note, ViewKey as PhoenixViewKey};
+use execution_core::{
+    transfer::phoenix::{
+        Note, SecretKey as PhoenixSecretKey, ViewKey as PhoenixViewKey,
+    },
+    BlsScalar,
+};
+
+/// Filter all notes that are owned by the given keys, mapped to their
+/// nullifiers.
+pub fn filter_map(
+    keys: impl AsRef<[PhoenixSecretKey]>,
+    notes: impl AsRef<[Note]>,
+) -> BTreeMap<BlsScalar, Note> {
+    let mut notes_map = BTreeMap::new();
+
+    notes.as_ref().iter().for_each(|note| {
+        for sk in keys.as_ref() {
+            if sk.owns(note.stealth_address()) {
+                let nullifier = note.gen_nullifier(sk);
+                notes_map.insert(nullifier, note.clone());
+            }
+        }
+    });
+
+    notes_map
+}
```

Comment:
Similar to the function to generate keys, I would probably avoid to create an external mutable variable and having everything self-contained in a `fold`:

```rust
pub fn find_a_better_name_here (
    keys: impl AsRef<[PhoenixSecretKey]>,
    notes: impl AsRef<[Note]>,
) -> BTreeMap<BlsScalar, Note> {
    notes
        .as_ref()
        .iter()
        .fold(BTreeMap::new(), |mut notes_map, note| {
            for sk in keys.as_ref() {
                if sk.owns(note.stealth_address()) {
                    let nullifier = note.gen_nullifier(sk);
                    notes_map.insert(nullifier, note.clone());
                }
            }
            notes_map
        })
}
```